server/src/main/java/org/elasticsearch/search/aggregations/TotalBucketCardinality.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations;
+
+import org.elasticsearch.search.aggregations.bucket.BucketsAggregator;
+import org.elasticsearch.search.aggregations.metrics.MetricsAggregator;
+
+/**
+ * Rough measure of how many buckets this {@link Aggregator} will collect
+ * used to pick data structures used during collection. Just "none", "one",
+ * and "many".
+ * <p>
+ * Unlike {@link AggregationBuilder.BucketCardinality} this is influenced
+ * by the number of buckets that the parent aggregation collect.
+ */
+public enum TotalBucketCardinality {
```

Comment:
A better name for this might be `CardinalityEstimate` or `CardinalityUpperBound`.

Reply:
I think removing "bucket" from the name would prevent some confusion. And it'd force it into something as simple as an upper bound on the `owningBucketOrd` passed to the collect method.

Reply:
That isn't what it is now, but it could be!

Reply:
Thinking about it again, there are really two cardinalities to measure and they are both useful.
1. The `Aggregator`'s cardinality. This is the thing I keep calling `parentCardinatlity` in the PR. It is an upper bound for the value of `owningBucketOrd` that will be passed to collect.
2. The `Aggregator`'s bucket's cardinality. This is the thing I was calling `bucketCardinality` in the PR. Confusingly, it is only a thing in `BucketsAggregation`s.

Also confusingly, there are a couple of relationships between these things:
1. An `Aggregator` can calculate its bucket's cardinality - sometimes exactly like `range` and sometimes not like `terms`.
2. An `Aggregator`'s bucket's cardinality *is* the cardinality of is sub-aggregations.

This PR, as it stands now, mixes those two things up and gives them funny names. I'd like to rework it.


To be clear: this is useful because knowing an `Aggregator`'s cardinality with some precision would let us pick smarter aggregation implementations. This PR only gives three levels of precision: `NONE`, `ONE`, and `MANY`. But we can do better in a follow up when we're sure how we're going to use it. Right now it is useful to have the distinction between `NONE` and `ONE` and `MANY`. Or, rather, it *would* be useful if I got the names straight. Right now I'm fairly sure it is wrong.